p4-16/spec/P4-16-spec.mdk

```diff
@@ -1496,12 +1495,21 @@ Each parameter may be labeled with a direction:
   call, the value of the parameter is copied to the corresponding storage location
   for that l-value.
 - `inout` parameters are both `in` and `out`. An argument passed as
-    an `inout` parameter must be an l-value.
-- No direction indicates that value of parameter is either:
-	- a compile-time known value
-	- an action parameter that can only be set by the control plane
-	- an action parameter that can be set directly by another calling
-      action; in this case it behaves like an `in` parameter
+  an `inout` parameter must be an l-value.
+- The meaning of parameters with no direction depends upon the kind of
+  entity the parameter is for:
+  - For anything other than an action, e.g. a control, parser, or
+    function, a directionless parameter means that the value supplied
+    as an argument in a call must be a compile-time known value.
+  - For an action, a directionless parameter indicates:
+    - If the action appears in an `actions` table property (Section
+      [#sec-table-action-list]), then the parameter's value must be
+      provided by the control plane software when an entry is added to
+      the table that uses that action.
+    - If the action is called in some other way, e.g. directly from
+      within a control, then the caller must provide a value, and the
+      call behaves the same as if the direction were `in`.  Such an
```

Comment:
Actually I don't think that this is true, they should still be a compile-time constant.

Reply:
I do not see anything else in the spec that implies this about actions with directionless parameters.  p4c as of a very recent version accepts programs that call such actions with run-time variable expressions as action arguments.  I can compile attached try3.p4.txt program without error using this version (I may need to attach that program to a different comment -- Github is complaining about attaching it to this comment):
```
$ p4c --version
p4c 1.2.0 (SHA: 98a890767 BUILD: DEBUG)
```

Reply:
I understand that this might be a case where p4c is doing more than is allowed by the spec.  The test program and p4c behavior is background that perhaps this new proposed spec text might be correct as written.  I do not know what other P4 compilers do with this yet.